ENH added recipe for MySQL server

beckermr opened the pull request:
This is a recipe for a full MySQL server installation. It is based on the homebrew package of the same name.

Some notes:
- I went ahead and used the cmake options to use bundled SSL, readline, editline and boost. This makes the packaging easier and more reliable. 
- I am running the shipped tests during the build. They are a lot of data and so running them right away means we can delete the directories.

Possible To Do:
- [x] ~~start server upon activation (following the toolchain package)~~

@druvus for viz

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/mysql```) and found it was in an excellent condition.

beckermr commented:
I think a separate conda package that you can add to automatically start the server upon activating the env makes more sense the adding that feature here.

dougalsutherland commented on recipes/mysql/build.sh:
```diff
@@ -0,0 +1,73 @@
+make install
+
+# we will run this test now and then delete the directory
+# there is no reason to ship the test dir and it is big
```
You could use [`test/source_files`](https://conda.io/docs/user-guide/tasks/build-packages/define-metadata.html#source-files) to let this happen in the test step. But testing here is probably fine too, especially since you're still testing that the commands run okay.

dougalsutherland commented on recipes/mysql/meta.yaml:
```diff
@@ -0,0 +1,48 @@
+    - toolchain
+    - cmake
+
+  run:
```
Drop this line if there's nothing there.

dougalsutherland replied:
`openssl` should be [pinned](https://github.com/conda-forge/conda-forge.github.io/blob/master/scripts/pin_the_slow_way.py#L81) at `openssl 1.0.*` in both build and run.

`libedit` should probably be pinned as `libedit 3.1` for now; investigating further.

dougalsutherland commented on recipes/mysql/meta.yaml:
```diff
@@ -0,0 +1,48 @@
+
+about:
+  home: https://www.mysql.com/
+  license: GPL
```
`GPL-2.0`

dougalsutherland commented on recipes/mysql/post-link.sh:
```diff
@@ -0,0 +1,40 @@
+
+    mysql_secure_installation
+
+2. By defuault, MySQL will only allow connections from localhost.
```
"default"

dougalsutherland commented on recipes/mysql/post-link.sh:
```diff
@@ -0,0 +1,40 @@
+fi
+
+# Make a symlink to script to start the server directly.
+ln -s ${PREFIX}/mysql/support-files/mysql.server ${PREFIX}/bin/mysql.server
```
Could this be handled by just putting this `ln` line in `build.sh`, so the symlink is just one of the files in the package and doesn't need to be handled by pre/post scripts?

dougalsutherland commented:
It's a little weird that databases are created by `post-link` but not removed by `pre-unlink`. I don't think they *should* be (that's a good way to lose all your data when you upgrade the package), but it maybe highlights that perhaps the database should live somewhere else? ...I don't know where else, though.

beckermr commented:
OK @dougalsutherland. I think I addressed all of your comments and got the OSX build problem.

Regarding where to put databases, I can see two approaches.

1. Do no initialization whatsoever and let users put their databases where they please. 
2. Set things up so each conda env has its own database that is made when the package is installed (what is happening now). 

The second option seems more in the spirit of conda, but I can understand a strong argument for the first. Namely, we package software, not databases. 

FWIW, homebrew takes the approach of installing a single global database. I have avoided that here since it didn't seem advisable given how conda is used (at least by me).

beckermr commented:
A third way would be to ship a second package that instantiates a database and starts it automatically when the env is activated.

dougalsutherland commented:
This question maybe wants input from @conda-forge/core.

Will `conda remove --all -n env` work properly if stray files are left behind? I think it'll complain and not delete them, right?

beckermr commented:
Worked fine for me.

```bash
(test) $ conda install mysql --use-local
Fetching package metadata ...............
Solving package specifications: .

Package plan for installation in environment /Users/mbecker/miniconda3/envs/test:

The following NEW packages will be INSTALLED:

    mysql: 5.7.20-0 local

Proceed ([y]/n)? y

MySQL Installation Notes:

1. WARNING: MySQL has been installed without a root password. To secure the
   installation, run:

    mysql_secure_installation

2. By defuault, MySQL will only allow connections from localhost.

3. To start [stop] the server, run:

    mysql.server start [stop]

4. The MySQL data directory is located within your Conda environment. This
   means that if your Conda environment is deleted, any databases will be
   deleted too!

Initialized MySQL server data directory.

(test) $ ls ~/miniconda3/envs/test/
bin/        conda-meta/ etc/        include/    lib/        mysql/      share/      ssl/        
(test) $ source deactivate
$ conda remove --all -n test

Package plan for package removal in environment /Users/mbecker/miniconda3/envs/test:

The following packages will be REMOVED:

    ca-certificates: 2017.7.27.1-0      conda-forge
    certifi:         2017.7.27.1-py36_0 conda-forge
    mysql:           5.7.20-0           local      
    ncurses:         5.9-10             conda-forge
    openssl:         1.0.2l-0           conda-forge
    pip:             9.0.1-py36_0       conda-forge
    python:          3.6.3-0            conda-forge
    readline:        6.2-0              conda-forge
    setuptools:      36.6.0-py36_1      conda-forge
    sqlite:          3.13.0-1           conda-forge
    tk:              8.5.19-2           conda-forge
    wheel:           0.30.0-py_1        conda-forge
    xz:              5.2.3-0            conda-forge
    zlib:            1.2.8-3            conda-forge

Proceed ([y]/n)? y

$ ls  /Users/mbecker/miniconda3/envs/test
ls: /Users/mbecker/miniconda3/envs/test: No such file or directory
```

beckermr commented:
If we go with option 2 or the third thing I posted, these docs look relevant.

https://dev.mysql.com/doc/refman/5.7/en/multiple-servers.html

dougalsutherland commented on recipes/mysql/post-link.sh:
```diff
@@ -0,0 +1,64 @@
+   deleted too!" >> ${PREFIX}/.messages.txt
+
+# the default port is generated by hasing the conda env name
+if [ -x "$(command -v md5sum)" ]
```
If you did this in python, with `hashlib`, you wouldn't have to do checks like this / it'd still work if whatever person's machine doesn't have either `md5` or `md5sum` on their `PATH`.

dougalsutherland replied:
Though actually, python isn't otherwise a runtime dep (and shouldn't be)...is there some simple in-bash way to do some version of this?

beckermr replied:
We could use cksum which is the same on both OSX and linux and might be more portable.

jakirkham commented:
Why not include a `pre-unlink` script to cleanup the unneeded files?

beckermr commented:
@jakirkham which files?

jakirkham commented:
IDK haven't read anything here yet. Sorry. Just responding to this [comment]( https://github.com/conda-forge/staged-recipes/pull/4224#issuecomment-339349053 ).

jakirkham commented:
Presumably anything that `post-link` is adding that was not in the package.

beckermr commented:
Ohhh yeah. We don't want to do that. We would delete people's databases. :(

jakirkham commented:
So that raises a fair question then. Should they be living in a `conda` environment if they are not transient?

beckermr commented:
Yep. This is the crux of the issue that we are debating. :)

I see three options:

1. Only ship mysql compiled and let people decided what to do themselves for making a database.
2. Ship the code such that the database install is completely local to each conda env.
3. Ship a separate package to make a local database in the conda env if people want to do that.
4. Ship a separate package to install mysql globally.

jakirkham commented:
Isn't it pretty easy to create a database with MySQL or are there some extra steps that I'm not thinking of?

beckermr commented:
It's not so bad. I am happy to not include that in the package. I have it here because homebrew did that and it seemed nice.

beckermr commented:
You just have to run the command in the post-link.sh script.

beckermr commented:
OK @jakirkham and @dougalsutherland. I think given our discussion we should remove the database install. We can ship a separate package that installs and sets up a database.

beckermr commented:
The OSX build died because the log file was too long. :) I tried to turn down the cmake verbosity.

beckermr commented:
We are soooooo close.

beckermr commented:
Alright @dougalsutherland and @jakirkham. It passed after a lot of OSX versus linux munging of the build! Yay!

Are we ready to merge?

jakirkham commented on recipes/mysql/meta.yaml:
```diff
@@ -0,0 +1,47 @@
+build:
+  number: 0
+  skip: True  # [win]
+  script: build.sh
```
This happens automatically with `conda-build`. So should not be needed.

jakirkham commented on recipes/mysql/build.sh:
```diff
@@ -0,0 +1,72 @@
+    rm -rf ${mysql_temp_dir}
+    cd -
+fi
+rm -rf ${PREFIX}/mysql-test
```
Would it make more sense to include this chunk in the `test` phase of the build instead? Could add to `run_test.sh` then `conda-build` will run this for us.

jakirkham replied:
Also should we be requiring `perl` if it is needed for testing?

dougalsutherland replied:
I think the idea was to not include `mysql-test` in the shipped package, since it's big. Can probably get around that with [`test/source_files`](https://conda.io/docs/user-guide/tasks/build-packages/define-metadata.html#source-files), though. (In that case, would also be able to ensure `perl` is present with `test/requires`.)

jakirkham replied:
Right, that's what I'm thinking as well.

beckermr replied:
These tests do not appear to be made to run outside the build directory. We can require perl during the build and run them right away.

jakirkham replied:
So we can copy whole directories with `test/source_files` if that helps. If it doesn't, we should start requiring `perl` in `build`. Either way please pin `perl` like [so]( https://github.com/conda-forge/openblas-feedstock/blob/4a65f26c18cfa4dea9b367016855d1d1258d4384/recipe/meta.yaml#L26 ).

beckermr replied:
I tried that. Adding perl to the build.

jakirkham commented on recipes/mysql/build.sh:
```diff
@@ -0,0 +1,72 @@
+        mkdir -p `dirname ${pth}` || :
+        ln -s `which cpp` ${pth} || :
+    fi
+done
```
This feels a little hacky/iffy to me. While it doesn't hurt the CIs, wouldn't feel comfortable running this on my own machine. Can you please point to where this is needed?

beckermr replied:
So on the linux build CI, cmake cannot find cpp. See this job https://circleci.com/gh/conda-forge/staged-recipes/20853?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

I have pasted the error here:

```
...
cd /staged-recipes/build_artefacts/mysql_1508768565518/work/mysql-5.7.20/build/rapid/plugin/group_replication/xdr_gen && rpcgen -C -h -o /staged-recipes/build_artefacts/mysql_1508768565518/work/mysql-5.7.20/build/rapid/plugin/group_replication/xdr_gen/xcom_vp.h xcom_vp.x
cannot find any C preprocessor (cpp)
rpcgen: C preprocessor failed with exit code 1
...
```

jakirkham replied:
So this came up in Arch Linux and they proposed a fix for [`rpcgen` calls]( https://bbs.archlinux.org/viewtopic.php?pid=1126049#p1126049 ). Guessing this doesn't happen on macOS.

beckermr replied:
Ahhh cool! I will try that.

jakirkham replied:
Might try doing something like `CPP_ROOT="dirname $(which cpp)"` and then replacing the `rpcgen` calls to be something like `rpcgen -Y $CPP_ROOT`. Would need double quotes instead single quotes for `sed`.

jakirkham commented on recipes/mysql/build.sh:
```diff
@@ -0,0 +1,49 @@
+
+# make sure we can find cpp on the linux CI service
+CPP_ROOT=`dirname $(which cpp)`
+sed -i'' -e "s|= rpcgen|= rpcgen -Y ${CPP_ROOT}|" */Makefile
```
What is the reason for the `*/` preceding `Makefile`?

beckermr replied:
This code is outdated. Sry.

jakirkham commented on recipes/mysql/build.sh:
```diff
@@ -0,0 +1,61 @@
+    -DWITH_SSL=bundled \
+    -DWITH_EDITLINE=bundled \
+    -DWITH_READLINE=bundled \
+    -DWITH_BOOST=boost \
```
Is it possible not to bundle these things? We include them in `conda-forge`. So it would be nice to depend on them instead.

beckermr replied:
Bundling makes shipping this all easier. I tried to not to bundle and it was hard.

jakirkham replied:
Could you please provide some more details on what you mean? Ideally shipping will just be installing with `conda`. What am I missing?

beckermr replied:
Sorry. I meant that compiling mysql against the conda shipped packages is a pain.

dougalsutherland replied:
Did you have the `toolchain` package as a build requirement and pass `-DCMAKE_PREFIX_PATH="$PREFIX"` to `cmake`? That's usually enough....

beckermr replied:
I had the package but I did not know about the extra cmake flag. I am happy to try but I am still curious why it is preferable to use the conda forge package. Letting MySQL bundle the code it needs eliminates dependencies and leads to a more robust build. 

Thoughts?

dougalsutherland replied:
It also introduces extra copies of the libraries on the system, makes the package bigger, and could conceivably lead to problems (maybe? not sure how mysql does it) if someone has a program that links to mysql and also a different version of readline.

jakirkham replied:
Everything @dougalsutherland said is correct. Would add more generally that the purpose of adding a package to `conda-forge` is to leverage a well maintained ecosystem of dependencies to lessen the burden of any one maintainer.

For instance, if `mysql` builds `boost` (as an example) in some suboptimal way, then you are on the hook as the `mysql` maintainer for fixing that bundled version somehow. However if you use `conda-forge`'s `boost`, you can lean on the `boost` maintainers here for help. This lessens the amount of working that you are having to do maintaining this package and helps shift the knowledge around in a more efficient way. Namely `boost` maintainers hone their expertise and share it with everyone needing help with `boost`. IOW you just need to focus on making `mysql` work right.

dougalsutherland replied:
Oh, one more benefit: builds should be faster and thus use up less CI resources when you don't also have to build `boost`/`readline`/`editline` with each build of `mysql`.

beckermr replied:
FWIW, the bundled libs don't ship extra copies. They must get compiled in or something.

```
$ ls -lah *
-rwxr-xr-x  1 mbecker  staff   4.1M Nov  1 11:00 libmysqlclient.20.dylib
-rw-r--r--  1 mbecker  staff   5.0M Nov  1 10:59 libmysqlclient.a
lrwxr-xr-x  1 mbecker  staff    23B Nov  1 11:00 libmysqlclient.dylib -> libmysqlclient.20.dylib
-rw-r--r--  1 mbecker  staff    35M Nov  1 10:59 libmysqld.a
-rw-r--r--  1 mbecker  staff   9.4K Nov  1 10:59 libmysqlservices.a

pkgconfig:
total 8
drwxr-xr-x  3 mbecker  staff   102B Nov  1 11:27 .
drwxr-xr-x  9 mbecker  staff   306B Nov  1 11:27 ..
-rw-r--r--  1 mbecker  staff   993B Nov  1 11:01 mysqlclient.pc

plugin:
total 7416
drwxr-xr-x  50 mbecker  staff   1.7K Nov  1 11:27 .
drwxr-xr-x   9 mbecker  staff   306B Nov  1 11:27 ..
-rwxr-xr-x   1 mbecker  staff    19K Nov  1 11:01 adt_null.so
-rwxr-xr-x   1 mbecker  staff    13K Nov  1 11:00 auth.so
-rwxr-xr-x   1 mbecker  staff    13K Nov  1 11:01 auth_test_plugin.so
-rwxr-xr-x   1 mbecker  staff    46K Nov  1 11:00 authentication_ldap_sasl_client.so
-rwxr-xr-x   1 mbecker  staff    47K Nov  1 11:01 connection_control.so
-rw-r--r--   1 mbecker  staff   227B Nov  1 10:31 daemon_example.ini
-rwxr-xr-x   1 mbecker  staff   969K Nov  1 11:00 group_replication.so
-rwxr-xr-x   1 mbecker  staff    31K Nov  1 11:00 ha_example.so
-rwxr-xr-x   1 mbecker  staff    89K Nov  1 11:01 keyring_file.so
-rwxr-xr-x   1 mbecker  staff    20K Nov  1 11:01 keyring_udf.so
-rwxr-xr-x   1 mbecker  staff    14K Nov  1 11:00 libdaemon_example.so
-rwxr-xr-x   1 mbecker  staff   8.5K Nov  1 11:00 libtest_framework.so
-rwxr-xr-x   1 mbecker  staff    14K Nov  1 11:00 libtest_services.so
-rwxr-xr-x   1 mbecker  staff    14K Nov  1 11:01 libtest_services_threaded.so
-rwxr-xr-x   1 mbecker  staff    19K Nov  1 11:01 libtest_session_detach.so
-rwxr-xr-x   1 mbecker  staff    14K Nov  1 11:00 libtest_session_in_thd.so
-rwxr-xr-x   1 mbecker  staff    31K Nov  1 11:00 libtest_session_info.so
-rwxr-xr-x   1 mbecker  staff    23K Nov  1 11:00 libtest_sql_2_sessions.so
-rwxr-xr-x   1 mbecker  staff    27K Nov  1 11:00 libtest_sql_all_col_types.so
-rwxr-xr-x   1 mbecker  staff    29K Nov  1 11:01 libtest_sql_cmds_1.so
-rwxr-xr-x   1 mbecker  staff    27K Nov  1 11:00 libtest_sql_commit.so
-rwxr-xr-x   1 mbecker  staff    28K Nov  1 11:00 libtest_sql_complex.so
-rwxr-xr-x   1 mbecker  staff    23K Nov  1 11:00 libtest_sql_errors.so
-rwxr-xr-x   1 mbecker  staff    27K Nov  1 11:00 libtest_sql_lock.so
-rwxr-xr-x   1 mbecker  staff    23K Nov  1 11:00 libtest_sql_processlist.so
-rwxr-xr-x   1 mbecker  staff    24K Nov  1 11:00 libtest_sql_replication.so
-rwxr-xr-x   1 mbecker  staff    19K Nov  1 11:00 libtest_sql_shutdown.so
-rwxr-xr-x   1 mbecker  staff    28K Nov  1 11:00 libtest_sql_sqlmode.so
-rwxr-xr-x   1 mbecker  staff    28K Nov  1 11:00 libtest_sql_stored_procedures_functions.so
-rwxr-xr-x   1 mbecker  staff    28K Nov  1 11:00 libtest_sql_views_triggers.so
-rwxr-xr-x   1 mbecker  staff    14K Nov  1 11:00 libtest_x_sessions_deinit.so
-rwxr-xr-x   1 mbecker  staff    14K Nov  1 11:00 libtest_x_sessions_init.so
-rwxr-xr-x   1 mbecker  staff    13K Nov  1 11:00 locking_service.so
-rwxr-xr-x   1 mbecker  staff    13K Nov  1 11:00 mypluglib.so
-rwxr-xr-x   1 mbecker  staff   8.6K Nov  1 11:00 mysql_no_login.so
-rwxr-xr-x   1 mbecker  staff   1.5M Nov  1 11:01 mysqlx.so
-rwxr-xr-x   1 mbecker  staff   8.3K Nov  1 11:01 qa_auth_client.so
-rwxr-xr-x   1 mbecker  staff    13K Nov  1 11:01 qa_auth_interface.so
-rwxr-xr-x   1 mbecker  staff   8.7K Nov  1 11:00 qa_auth_server.so
-rwxr-xr-x   1 mbecker  staff    26K Nov  1 11:00 replication_observers_example_plugin.so
-rwxr-xr-x   1 mbecker  staff    13K Nov  1 11:01 rewrite_example.so
-rwxr-xr-x   1 mbecker  staff    48K Nov  1 11:00 rewriter.so
-rwxr-xr-x   1 mbecker  staff    56K Nov  1 11:00 semisync_master.so
-rwxr-xr-x   1 mbecker  staff    16K Nov  1 11:00 semisync_slave.so
-rwxr-xr-x   1 mbecker  staff    13K Nov  1 11:00 test_security_context.so
-rwxr-xr-x   1 mbecker  staff   8.7K Nov  1 11:00 test_udf_services.so
-rwxr-xr-x   1 mbecker  staff    43K Nov  1 11:00 validate_password.so
-rwxr-xr-x   1 mbecker  staff    47K Nov  1 11:00 version_token.so
```

beckermr replied:
Sounds good all. I will work to get it to compile against the shipped packages.

dougalsutherland replied:
Yeah, that sounds like the best thing to do for boost.

beckermr commented:
Ok. I think I got everything. Ready to merge?

beckermr commented:
Whoops. Missed the comments on pinning libedit and OpenSSL. Will do

jakirkham commented on recipes/mysql/meta.yaml:
```diff
@@ -0,0 +1,53 @@
+    - toolchain
+    - cmake
+    - perl 5.22.2.1
+    - libedit
```
So we typically use `readline` for this sort of thing. Admittedly we added `libedit` recently, but more as a crutch for a `conda` solver tar pit. Any thoughts on using `readline`?

beckermr commented:
Yeah so this is weird. I had ‘WITH_READLINE’ in the cmake invocation. Then I noticed in a build that cmake reported that it did not use that option, which was weird. So I double checked the MySQL docs and WITH_READLINE does not appear to be a valid option. Fun. 

So given all of that, unless the libedit headers are included in the readline package, we should stick with libedit.

jakirkham commented:
np. Mentioned mainly as their APIs are basically compatible and most things can use either. That said, having not built MySQL from source, am trusting your judgement here about what works well or not.

beckermr commented:
Ok @dougalsutherland and @jakirkham. Are we good to go on this?

beckermr commented:
I think our conclusion on the tests was that they didn’t run well outside of the build and so we are leaving them where they are. With the perl dep they now run everywhere.

jakirkham commented:
Yeah had no strong feelings on where the tests ran. Only that they did run and had their dependencies met, which I believe we addressed.

Where did we land on the `libedit` pinning?

jakirkham commented on recipes/mysql/meta.yaml:
```diff
@@ -0,0 +1,53 @@
+    - openssl 1.0.*
+
+  run:
+    - libedit
```
If you are going with `3.1`, would reuse the same thing here too.

jakirkham commented on recipes/mysql/meta.yaml:
```diff
@@ -0,0 +1,53 @@
+
+  run:
+    - libedit
+    - openssl
```
Also this should be `openssl 1.0.*`.

jakirkham commented:
Minor comments on the pinnings. Should be easy fixes. If anyone else has insights into `libedit`'s API/ABI compatibility, would be happy to hear them. Otherwise pinning to the exact version we have in `conda-forge` seems safe and allows for this to be changed later on as needed. Once that is settled, we should merge.

beckermr commented:
Good catch on pinning the runtime deps too @jakirkham! Hopefully the tests will run overnight and we can merge tomorrow!

beckermr commented:
Looks like this is good to go!

jakirkham commented:
This looks good.

Though we discovered a [surprising thing]( https://github.com/conda-forge/conda-forge.github.io/issues/478#issuecomment-341704800 ) about `libedit`. To be clear, this isn't on you to fix, it's on us. That said, feel free to wade into the discussion if you wish.

beckermr commented:
Wow. Mmmmk. Are blocked on this then @jakirkham?

dougalsutherland commented:
I don't think it should take long to resolve – https://github.com/conda-forge/libedit-feedstock/pull/1 – but let's see if we can do that quickly first.

dougalsutherland commented:
https://github.com/conda-forge/libedit-feedstock/pull/1 merged, so I went ahead and updated the `libedit` pin and then merged. Thanks @beckermr!
